File: crates/ffi/lang/python/examples/example_project/emit_manifest.py
"""
Emit a Daedalus manifest from Python nodes.

Run (from repo root):
  python crates/ffi/lang/python/examples/example_project/emit_manifest.py /tmp/example_py.manifest.json

This writes a manifest and copies any shader files next to it.
"""

from __future__ import annotations

import sys
import os
import tempfile
from pathlib import Path

# Import the in-repo SDK directly.
REPO_ROOT = Path(__file__).resolve().parents[6]
PY_SDK = REPO_ROOT / "crates" / "ffi" / "lang" / "python"
sys.path.insert(0, PY_SDK.as_posix())

from daedalus_py import Plugin  # noqa: E402

from nodes import register_all  # noqa: E402

def copy_shaders(out_dir: Path) -> None:
    src_dir = Path(__file__).with_name("shaders")
    dst_dir = out_dir / "shaders"
    dst_dir.mkdir(parents=True, exist_ok=True)
    for p in src_dir.glob("*.wgsl"):
        (dst_dir / p.name).write_text(p.read_text(encoding="utf-8"), encoding="utf-8")


def main() -> None:
    out = Path(tempfile.gettempdir()) / f"example_py_{os.getpid()}.manifest.json"
    if len(sys.argv) > 1:
        out = Path(sys.argv[1])
    out_dir = out.parent

    copy_shaders(out_dir)

    plugin = Plugin(name="example_py", version="0.1.0", description="Python example project")
    register_all(plugin)

    # Point nodes at a stable module file (no codegen step).
    nodes_path = Path(__file__).with_name("nodes.py").resolve().as_posix()
    for n in plugin.nodes:
        n.py_function = n.id.split(":")[-1]
        n.py_path = nodes_path

    # Optional: add a shader to a node by setting n.shader = {...} or using helper APIs.
    # Keep it file-backed: src_path should point to a real file next to the manifest.
    #
    # Example (not registered by default):
    # some_node.shader = {
    #   "src_path": "shaders/invert.wgsl",
    #   "entry": "main",
    #   "name": "invert",
    #   "invocations": [1, 1, 1],
    #   "bindings": [
    #     {"binding": 0, "kind": "storage_buffer", "access": "read_write", "readback": True, "to_port": "out", "size_bytes": 4},
    #   ],
    # }

    plugin.emit_manifest(out)
    print(out.as_posix())


if __name__ == "__main__":
    main()

File: crates/ffi/lang/node/examples/example_project/js/emit_manifest.mjs
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import process from "node:process";

// Import the in-repo SDK directly.
const repoRoot = path.resolve(path.dirname(new URL(import.meta.url).pathname), "../../../../../../..");
const sdk = path.join(repoRoot, "crates", "ffi", "lang", "node", "daedalus_node", "index.mjs");
const { Plugin, NodeDef, t } = await import(`file://${sdk}`);

function copyShaders(outDir) {
  const srcDir = path.join(path.dirname(new URL(import.meta.url).pathname), "shaders");
  const dstDir = path.join(outDir, "shaders");
  fs.mkdirSync(dstDir, { recursive: true });
  for (const name of fs.readdirSync(srcDir)) {
    if (!name.endsWith(".wgsl")) continue;
    fs.copyFileSync(path.join(srcDir, name), path.join(dstDir, name));
  }
}

const out = process.argv[2]
  ? path.resolve(process.argv[2])
  : path.join(os.tmpdir(), `example_node_${process.pid}.manifest.json`);
fs.mkdirSync(path.dirname(out), { recursive: true });
copyShaders(path.dirname(out));

const plugin = new Plugin({ name: "example_node", version: "0.1.1", description: "Node example project" });

// Point to the local nodes module sitting next to this emitter.
const nodesPath = path.join(path.dirname(new URL(import.meta.url).pathname), "nodes.mjs");

plugin.register(
  new NodeDef({
    id: "example_node:add",
    js_path: nodesPath,
    js_function: "add",
    inputs: [
      { name: "a", ty: t.int() },
      { name: "b", ty: t.int() },
    ],
    outputs: [{ name: "out", ty: t.int() }],
    metadata: { lang: "node", kind: "stateless" },
  }),
);

plugin.register(
  new NodeDef({
    id: "example_node:split",
    js_path: nodesPath,
    js_function: "split",
    inputs: [{ name: "value", ty: t.int() }],
    outputs: [
      { name: "out0", ty: t.int() },
      { name: "out1", ty: t.int() },
    ],
    metadata: { lang: "node", kind: "multi_output" },
  }),
);

plugin.register(
  new NodeDef({
    id: "example_node:counter",
    js_path: nodesPath,
    js_function: "counter",
    inputs: [{ name: "inc", ty: t.int() }],
    outputs: [{ name: "out", ty: t.int() }],
    stateful: true,
    state: { start: 0 },
    metadata: { lang: "node", kind: "stateful" },
  }),
);

// Optional: shader wiring example (leave disabled by default).
// Keep it file-backed: src_path should be relative to the manifest directory.
//
// plugin.register(new NodeDef({
//   id: "example_node:invert_first_u32",
//   js_path: nodesPath,
//   js_function: "add", // placeholder: the node won't call JS when shader is present
//   inputs: [{ name: "in", ty: t.bytes() }],
//   outputs: [{ name: "out", ty: t.bytes() }],
//   shader: {
//     src_path: "shaders/invert.wgsl",
//     entry: "main",
//     name: "invert",
//     invocations: [1, 1, 1],
//     bindings: [
//       { binding: 0, kind: "storage_buffer", access: "read_write", readback: true, to_port: "out", size_bytes: 4 },
//     ],
//   },
// }));

plugin.emitManifest(out);
process.stdout.write(out + "\n");
